content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,252 @@
+  stopifnot("`name` must be a character." = is.character(name))
+  paste("Hello", name)
+}
+
+say_hello(404)
+```
+
+This is clearly a really great improvement to the functionality of base R.
+However, we can see from this example that we could create the error message programmatically based on the contents of the test.
+Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
+This way, you don't have to repeat yourself which is generally a good aim [^3].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^3]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
+
+```{r}
+say_hello <- function(name) {
+  # Among other things, check_string() checks that we provide a 
+  # character object of length one
+  checkmate::assert_string(name)
+  paste("Hello", name)
+}
+```
+
+```{r, error = TRUE}
+say_hello(404)
+```
+
+```{r, error = TRUE}
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
+
+Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
+We will not get into the details of all of the available options here but below is a list of some of them, listed by decreasing number of reverse dependencies.
+If interested in understanding the various approaches to input taking the documentation for these package is a great place to start.
+
+-   [assertthat](https://github.com/hadley/assertthat)
+
+```{r, error = TRUE}
+assertthat::assert_that(is.character(1))
+```
+
+-   [assertr](https://docs.ropensci.org/assertr/)
+
+```{r, error = TRUE}
+library(magrittr)
+
+mtcars %>%
+  assertr::verify(nrow(.) < 10)
+```
+
+-   [assertive](https://bitbucket.org/richierocks/assertive)
+
+```{r, error = TRUE}
+assertive::assert_is_a_string(1)
+```
+
+-   [ensurer](https://github.com/smbache/ensurer)
+
+```{r, error = TRUE}
+ensure_square <- ensurer::ensures_that(NCOL(.) == NROW(.))
+
+ensure_square(matrix(1:20, 4, 5))
+```
+
+-   `vctrs::vec_assert()`
+
+```{r, error = TRUE}
+vctrs::vec_assert(c(1, 2), "character")
```

Comment:
ça rend bizarre dans la preview (locale, `hugodown::hugo_start()`), y a-t-il une option à changer quand on l'utilise dans R Markdown ?

Reply:
Maybe the `NO_COLOR` environment variable https://cli.r-lib.org/reference/faq.html